File: leetcode/python/000143.reorder-list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
	# Time complexity: O(n)
	# Space complexity: O(n)
	def reorderList(self, head: Optional[ListNode]) -> None:
		"""
		Do not return anything, modify head in-place instead.
		"""
		pt1, pt2 = head, head
		while pt2 and pt2.next:
			pt1 = pt1.next
			pt2 = pt2.next.next

		reversed_half = self.reverse_list(pt1.next)
		pt1.next = None

		pt1 = head
		pt2 = reversed_half
		while pt1 and pt2:
			temp = pt1.next
			pt1.next = pt2
			pt1 = temp
			temp = pt2.next
			pt2.next = pt1
			pt2 = temp

	def reverse_list(self, head: Optional[ListNode]) -> Optional[ListNode]:
		pt1 = None
		pt2 = head
		while pt2:
			temp = pt2.next
			pt2.next = pt1
			pt1 = pt2
			pt2 = temp
		return pt1

File: leetcode/typescript/000143.reorder-list.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

/**
 Do not return anything, modify head in-place instead.
 */
function reorderList(head: ListNode | null): void {
  let slow = head;
  let fast = head;

  while (fast && fast.next) {
    fast = fast.next.next;
    slow = slow!.next;
  }

  const reversedHead = reverse(slow ? slow.next : null);

  if (slow) {
    slow.next = null;
  }

  let curA = head;
  let curB = reversedHead;
  let dummyHead = new ListNode(0);
  let cur = dummyHead;

  while (curA && curB) {
    cur.next = curA;
    curA = curA.next;
    cur = cur.next;

    cur.next = curB;
    curB = curB.next;
    cur = cur.next;
  }

  cur.next = curA;
}
// Time complexity: O(n)
// Space complexity: O(1)

function reverse(head: ListNode | null): ListNode | null {
  let prev: ListNode | null = null;
  let cur = head;

  while (cur) {
    const temp = cur.next;
    cur.next = prev;
    prev = cur;
    cur = temp;
  }

  return prev;
}
